libplatsupport/src/plat/morello-qemu/serial.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2019, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+/* Mostly copy/paste from the HiKey plat.
+ * Should be moved to a common driver file for PL011 */
+
+#include <string.h>
+#include <stdlib.h>
+#include <platsupport/serial.h>
+#include "../../chardev.h"
+
+#define RHR_MASK                MASK(8)
+#define UARTDR                  0x000
+#define UARTFR                  0x018
+#define UARTIMSC                0x038
+#define UARTICR                 0x044
+#define PL011_UARTFR_TXFF       BIT(5)
+#define PL011_UARTFR_RXFE       BIT(4)
+
+#define REG_PTR(base, off)     ((volatile uint32_t *)((base) + (off)))
+
+int uart_getchar(ps_chardevice_t *d)
+{
+    int ch = EOF;
+
+    if ((*REG_PTR(d->vaddr, UARTFR) & PL011_UARTFR_RXFE) == 0) {
+        ch = *REG_PTR(d->vaddr, UARTDR) & RHR_MASK;
+    }
+    return ch;
+}
+
+int uart_putchar(ps_chardevice_t *d, int c)
+{
+    while ((*REG_PTR(d->vaddr, UARTFR) & PL011_UARTFR_TXFF) != 0);
+
+    *REG_PTR(d->vaddr, UARTDR) = c;
+    if (c == '\n' && (d->flags & SERIAL_AUTO_CR)) {
+        uart_putchar(d, '\r');
```

Comment:
This writes `\n\r`, but it should be `\r\n`

Reply:
I just copied it from `libplatsupport/src/plat/qemu-arm-virt/serial.c`

Reply:
It is broken there also, see 970c79da for a fix